botorch/acquisition/multi_step_lookahead.py

```diff
@@ -213,44 +247,284 @@ def _step(
         Xs=Xs[1:],
         samplers=samplers[1:],
-        value_function_cls=value_function_cls,
-        value_function_kwargs=value_function_kwargs,
+        valfunc_cls=valfunc_cls[1:],
+        valfunc_argfacs=valfunc_argfacs[1:],
+        inner_samplers=inner_samplers[1:],
         objective=objective,
+        sample_weights=sample_weights,
+        running_val=running_val,
+    )
+
+
+def _compute_stage_value(
+    model: Model,
+    valfunc_cls: Optional[Type[AcquisitionFunction]],
+    X: Tensor,
+    objective: AcquisitionObjective,
+    sample_weights: Tensor,
+    inner_sampler: Optional[MCSampler] = None,
+    arg_fac: Optional[TAcqfArgConstructor] = None,
+) -> Optional[Tensor]:
+    r"""Compute the stage value of a multi-step look-ahead policy.
+
+    Args:
+        model: A Model of appropriate batch size. Specifically, it must be possible to
+            evaluate the model's posterior at `Xs[0]`.
+        valfunc_cls: The acquisition function class to be used as the stage value
+            functions. If `None`, a zero stage value is assumed (returns `None`)
+        X: A tensor with shape `f_i x .... x f_1 x batch_shape x q_i x d` when called in
+            the `i`-th step.
+        objective: The AcquisitionObjective under which the model output is evaluated.
+        sample_weights: A tensor of shape `f_i x .... x f_1 x batch_shape` when called
+            in the `i`-th step by which to weight the stage value samples. Used in
+            conjunction with Gauss-Hermite integration or importance sampling.
+        inner_sampler: An `MCSampler` object to be used in the stage value function. Can
+            be `None` for analytic acquisition funcitons or when using the default
+            sampler of the acquisition function class.
+        arg_fac: A callable mapping a `Model` and the input tensor `X` to a dictionary
+            of kwargs for the stage value function constructor. If `None`, only the
+            standard `model`, `sampler` and `objective` kwargs will be used.
+
+    Returns:
+        A `batch_shape`-dim tensor of stage values, or `None` (=zero stage value).
+    """
+    if valfunc_cls is None:
+        return None
+    common_kwargs: Dict[str, Any] = {"model": model, "objective": objective}
+    if issubclass(valfunc_cls, MCAcquisitionFunction):
+        common_kwargs["sampler"] = inner_sampler
+    kwargs = arg_fac(model=model, X=X) if arg_fac is not None else {}
+    stage_val_func = valfunc_cls(**common_kwargs, **kwargs)
+    stage_val = stage_val_func(X=X)
+    # shape of stage_val is (inner_mc_samples) x f_k x ... x f_1 x batch_shape
+    if sample_weights.numel() > 0:
+        # if there are sample weights, we need to use them
+        # sample_weights must be of shape f_k x ... x f_1 x batch_shape
+        stage_val = stage_val * sample_weights
+    # we average across all dimensions except for the batch dimension
+    stage_val = stage_val.view(-1, stage_val.size(-1)).mean(dim=0)
```

Comment:
in case of GH sampler, .mean() should be .sum()?  \int f(x) w(x) dx = w_1 f(x_1) + w_2 f(x_2) + ...